confgenerator/confgenerator.go

```diff
@@ -278,6 +423,59 @@ func extractReceiverFactories(receivers map[string]*receiver) (map[string]*fileR
 	return fileReceiverFactories, syslogReceiverFactories, wineventlogReceiverFactories, nil
 }
 
+func generateOtelReceivers(hostmetricsReceiverFactories map[string]*hostmetricsReceiverFactory, pipelines map[string]*otelPipeline) ([]*otel.HostMetrics, map[string]string, error) {
+	hostMetricsList := []*otel.HostMetrics{}
+	receiveNameMap := make(map[string]string)
+	var pipelineIDs []string
+	for p := range pipelines {
+		pipelineIDs = append(pipelineIDs, p)
+	}
+	sort.Strings(pipelineIDs)
+	for _, pID := range pipelineIDs {
+		p := pipelines[pID]
+		for _, rID := range p.Receivers {
+			if h, ok := hostmetricsReceiverFactories[rID]; ok {
+				hostMetrics := otel.HostMetrics{
+					HostMetricsID:      rID,
+					CollectionInterval: h.CollectionInterval,
+				}
+				hostMetricsList = append(hostMetricsList, &hostMetrics)
+				receiveNameMap[rID] = "hostmetrics/" + rID
+				continue
+			}
+			return nil, nil, fmt.Errorf(`receiver %q of pipeline %q is not defined`, rID, pID)
+		}
+	}
+	return hostMetricsList, receiveNameMap, nil
+}
+
+func generateOtelExporters(exporters map[string]*otelExporter, pipelines map[string]*otelPipeline) ([]*otel.Stackdriver, map[string]string, error) {
+	stackdriverList := []*otel.Stackdriver{}
+	exportNameMap := make(map[string]string)
+	var pipelineIDs []string
+	for p := range pipelines {
+		pipelineIDs = append(pipelineIDs, p)
+	}
+	sort.Strings(pipelineIDs)
+	for _, pID := range pipelineIDs {
+		p := pipelines[pID]
+		for _, eID := range p.Exporters {
+			if _, ok := exporters[eID]; ok {
+				stackdriver := otel.Stackdriver{
+					StackdriverID: eID,
+					UserAgent:     "$USERAGENT",
+					Prefix:        "agent.googleapis.com/",
+				}
+				stackdriverList = append(stackdriverList, &stackdriver)
+				exportNameMap[eID] = "stackdriver/" + eID
+				continue
+			}
+			return nil, nil, fmt.Errorf(`exporter %q of pipeline %q is not defined`, eID, pID)
```

Comment:
Should this check be outside the inner for loop? (Same above.)

Reply:
i think this is right place. The logic resembles here in logging side: https://github.com/GoogleCloudPlatform/ops-agent/blob/master/confgenerator/confgenerator.go#L325

First, looping a list of pipeLines then for each pipeline,  for looping this pipeline's exporter IDs, trying to find this ID inside the exporters. If it's not found, return the error.

Reply:
Oh sorry, I misread this code, thanks for the clarification.

I think it might be clearer to invert this condition:

    if _, ok := exporters[eID]; !ok {
      return nil, nil, ...
    }
    stackdriver := otel.Stackdriver{...}
    ...

Reply:
Yes, this way is much clearer! Done